docs/setup/config.md
# Config file options

With the configuration file, you can control certain aspects of the Telescope instance's operation.
The `TELESCOPE_CONFIG_FILE` variable specifies the configuration file that is used.

Please check the [config.py](https://github.com/iamtelescope/telescope/blob/main/backend/telescope/config.py) to get a detailed view of the default values used in the configuration and how they are merged with the values from the configuration files.

```yaml
auth:
  providers:
    github:
      # Enable or disable login via GitHub (https://github.com/iamtelescope/telescope/blob/main/backend/base/settings.py#L143-L154)
      enabled: false
      # Add users to the specified group on login (the group must exist)
      default_group: null
      # client_id, key, and secret should be obtained from GitHub (https://docs.allauth.org/en/dev/socialaccount/providers/github.html)
      client_id: ''
      key: ''
      secret: ''
      # If specified, users must belong to one of these organizations to log in.
      organizations: []
  # If set to true, forces login page to redirect to github auth (only available when github auth provider is enabled)
  force_github_auth: false
  # If set to true, the user is always authenticated without any conditions
  enable_testing_auth: false
  # If enable_testing_auth is set true, defines the username of the logged-in user
  testing_auth_username: telescope

# These options are passed to Django settings
django:
  CSRF_TRUSTED_ORIGINS:
  - http://localhost:9898
  DATABASES:
    default:
      ENGINE: django.db.backends.sqlite3
      NAME: telescope-default-db.sqlite3
  DEBUG: false

# Passed to the Gunicorn web server instance (https://docs.gunicorn.org/en/latest/settings.html) (https://github.com/iamtelescope/telescope/blob/main/backend/app.py#L25)
gunicorn:
  bind: 127.0.0.1:9898
  max_requests: 50
  max_requests_jitter: 50
  timeout: 120
  workers: 8

# Section defines various per-user or system-wide constraints for Telescope.
limits:
  # Specifies the maximum number of saved views a single user can create. If set to `0`, the limit is disabled and users can create unlimited views.
  max_saved_views_per_user: 0
  
# Options provided to vuejs app after loading config via `/ui/v1/config` handler
frontend:
    show_github_url: true  # show or not github url on top panel
    github_url: ""
    show_docs_url: true  # show or not docs url on top panel
    docs_url: ""

# Logging configuration (https://github.com/iamtelescope/telescope/blob/main/backend/telescope/log.py#L59)
logging:
  # Options: default, dev, or json
  format: default
  # Setup for log levels of different handlers
  levels:
    all: DEBUG
    django: DEBUG
    django.request: DEBUG
    django.template: INFO
    django.utils.autoreload: INFO
    telescope: DEBUG
```


docs/changelog.md
# Changelog

## 2025.05.18
Version: **0.0.18**
Changes:
  - Implemented saved views
  - Small bugfixes & improvements

Related issues:
 - [ISSUE-3](https://github.com/iamtelescope/telescope/issues/3)

Notes:

{% note warning %}

This release requires database migrations to be run.

{% endnote %}

## 2025.04.20
Version: **0.0.17**
Changes:
  - Implemented token-based authentication (API tokens) as a foundation for future API access
  - Bugfixed data section border misalignment in the Explorer. Thanks to [meirdev](https://github.com/meirdev)!

Related issues:
 - [ISSUE-24](https://github.com/iamtelescope/telescope/issues/24)

Notes:

{% note warning %}

This release requires database migrations to be run.

{% endnote %}

## 2025.04.08
Version: **0.0.16**
Changes:
  - Added various options for SSL-based ClickHouse connections
  - Added the ability to disable authentication via the config file

Notes:

{% note warning %}

This release requires database migrations to be run.

{% endnote %}

## 2025.04.01
Version: **0.0.15**
Changes:
  - Added Docker as a new source kind
  - Redesigned source list
  - Added toggle to show/hide filters on the explorer page
  - Fixed bug in date range selector when using the graph
  - Fixed bug where the show/hide graph option wasn’t preserved on reload

Notes:

{% note warning %}

This release requires database migrations to be run.

{% endnote %}

## 2025.03.19
Version: **0.0.14**

Changes:
 - The source time field now supports UInt64, Int64, and timestamp ClickHouse types. Thanks to [Zerg1996](https://github.com/Zerg1996)!

Related issues:
 - [ISSUE-13](https://github.com/iamtelescope/telescope/issues/13)

## 2025.03.15
Version: **0.0.13**

Changes:
 - Added config option to force github authentication on login page

## 2025.03.11
Version: **0.0.12**
Changes:
 - Explorer UI update: query settings and chart settings have been moved into separate dropdown menus to improve usability and reduce interface clutter.
 - Added a placeholder for the raw query editor and made various style adjustments

## 2025.03.10 +
Version: **0.0.11**
Changes:
 - Added support for raw SQL `WHERE` statements

Related issues:
 - [ISSUE-5](https://github.com/iamtelescope/telescope/issues/5)

## 2025.03.10
Version: **0.0.10**
Changes:
 - Added support for custom fields in the "Group By" option on graphs.
 - Changed data loading behavior: now, when making a new request, the old data remains on the screen until the new data is fully loaded.
 - Bugfix for a mismatch between time values in graphs and data tables (caused by differences between UTC and non-UTC timestamps).

Related issues:
 - [ISSUE-4](https://github.com/iamtelescope/telescope/issues/4)

Notes:

{% note warning %}

This release requires database migrations to be run.

{% endnote %}

## 2025.02.27
Version: **0.0.9**

Changes:
 - Make the source severity field optional.

Related issues:
 - [ISSUE-6](https://github.com/iamtelescope/telescope/issues/6)


## 2025.02.24
Version: **0.0.8**

Changes:
 - Bugfix for non-UTC time fields in ClickHouse
 - Bugfix for specifying time with histogram range selection

## 2025.02.19
Version: **0.0.7**

Changes:
- Cosmetic login screen update.

## 2025.02.18
Version: **0.0.6**

Changes:
- Fixed monaco editor static files.

## 2025.02.17
Version: **0.0.5**

Changes:
- Fixed a bug inside fields parser.
- Updated modifiers docs.
- Added naive support for Map & Array ClickHouse types.
- Added new modifiers:
    - [href](./ui/explorer/fields.md#href), for dynamically creating HTML links.
    - [format](./ui/explorer/fields.md#format), for pretty formatting of SQL or JSON data.
    - [highlight](./ui/explorer/fields.md#highlight), for syntax highlighting of SQL or JSON data.

## 2025.02.13
Version: **0.0.4**

Changes:
- Fixed default JSON renderer to handle unserializable objects properly.

## 2025.02.13
Version: **0.0.3**

Changes:
- Added placeholders for fields and query inputs based on source fields.

## 2025.02.12
Version: **0.0.2**

Changes:
 - Initial release


backend/telescope/config.py
import os
import collections.abc

import yaml
import jsonschema


class ConfigValidationError(RuntimeError):
    def __init__(
        self,
        message,
        errors=None,
    ):
        self.message = message
        self.errors = errors or []

    def __str__(self):
        return self.message + ": " + ", ".join([f"{i[0]}: {i[1]}" for i in self.errors])

    def __repr__(self):
        return str(self)


JSONSchemaValidator = jsonschema.Draft7Validator


SCHEMA = {
    "type": "object",
    "properties": {
        "gunicorn": {
            "type": "object",
        },
        "auth": {
            "type": "object",
            "properties": {
                "providers": {
                    "type": "object",
                },
                "force_github_auth": {
                    "type": "boolean",
                },
                "enable_testing_auth": {
                    "type": "boolean",
                },
                "testing_auth_username": {
                    "type": "string",
                },
            },
        },
        "limits": {
            "max_saved_views_per_user": {
                "type": "integer",
            },
        },
        "django": {
            "type": "object",
            "properties": {
                "SECRET_KEY": {
                    "type": "string",
                },
            },
            "required": ["SECRET_KEY"],
            "additionalProperties": True,
        },
        "logging": {
            "type": "object",
        },
        "frontend": {
            "type": "object",
            "properties": {
                "github_url": {
                    "type": "string",
                },
                "show_github_url": {
                    "type": "boolean",
                },
                "docs_url": {
                    "type": "string",
                },
                "show_docs_url": {
                    "type": "boolean",
                },
            },
        },
        "additionalProperties": True,
    },
}


class YamlLoader(yaml.SafeLoader):
    @classmethod
    def init_constructors(cls):
        cls.add_constructor("!env", cls.env)

    def __init__(self, stream):
        super(YamlLoader, self).__init__(stream)

    def env(self, node):
        env_var = self.construct_scalar(node)
        env_value = os.getenv(env_var)
        return env_value


YamlLoader.init_constructors()


def validate(config, schema):
    errors = []
    errors_generator = JSONSchemaValidator(schema).iter_errors(config)
    for error in sorted(errors_generator, key=jsonschema.exceptions.relevance):
        path = ".".join(map(str, error.absolute_path))
        errors.append((path, error.message))

    if not errors:
        if (
            config["auth"]["force_github_auth"]
            and not config["auth"]["providers"]["github"]["enabled"]
        ):
            errors.append(
                (
                    "auth.force_github_auth",
                    "cannot be true if github provider is not enabled",
                )
            )

    if errors:
        raise ConfigValidationError(message="Malformed configuration", errors=errors)


def merge_dicts(first, second):
    def _update(orig, update):
        for key, value in update.items():
            if isinstance(value, collections.abc.Mapping):
                orig[key] = _update(orig.get(key, {}), value)
            else:
                orig[key] = update[key]
        return orig

    _result = {}
    _update(_result, first)
    _update(_result, second)
    return _result


def get_default_config():
    return {
        "gunicorn": {
            "bind": "127.0.0.1:9898",
            "workers": 8,
            "timeout": 120,
            "max_requests": 50,
            "max_requests_jitter": 50,
        },
        "django": {
            "CSRF_TRUSTED_ORIGINS": [
                "http://localhost:9898",
            ],
            "DEBUG": False,
            "DATABASES": {
                "default": {
                    "ENGINE": "django.db.backends.sqlite3",
                    "NAME": "telescope-default-db.sqlite3",
                },
            },
        },
        "limits": {
            "max_saved_views_per_user": 0,
        },
        "auth": {
            "providers": {
                "github": {
                    "enabled": False,
                    "key": "",
                    "organizations": [],
                    "default_group": None,
                },
            },
            "force_github_auth": False,
            "enable_testing_auth": False,
            "testing_auth_username": "telescope",
        },
        "frontend": {
            "github_url": "https://github.com/iamtelescope/telescope",
            "docs_url": "https://iamtelescope.github.io/telescope/docs",
            "show_docs_url": True,
        },
        "logging": {
            "format": "default",
            "levels": {
                "django": "DEBUG",
                "django.request": "DEBUG",
                "django.template": "INFO",
                "django.utils.autoreload": "INFO",
                "telescope": "DEBUG",
                "all": "DEBUG",
            },
        },
    }


def get_config():
    config_file = os.environ.get("TELESCOPE_CONFIG_FILE")
    config = {}
    default_config = get_default_config()

    if config_file:
        with open(config_file) as fd:
            config = yaml.load(fd, Loader=YamlLoader)
    config = merge_dicts(default_config, config)
    validate(config, SCHEMA)
    return config
